### ISSUE TITLE
"composer validate" exits with code 1 when composer.json is valid

### ISSUE THREAD

**Issue by sebastianbergmann** [composer/composer#7624]
```
$ git clone https://github.com/sebastianbergmann/phpunit.git
$ cd phpunit
```
```
$ composer install
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 26 installs, 0 updates, 0 removals
  - Installing theseer/tokenizer (1.1.0): Loading from cache
  - Installing sebastian/version (2.0.1): Loading from cache
  - Installing sebastian/environment (3.1.0): Loading from cache
  - Installing sebastian/code-unit-reverse-lookup (1.0.1): Loading from cache
  - Installing phpunit/php-token-stream (3.0.0): Loading from cache
  - Installing phpunit/php-text-template (1.2.1): Loading from cache
  - Installing phpunit/php-file-iterator (2.0.1): Loading from cache
  - Installing phpunit/php-code-coverage (6.0.7): Loading from cache
  - Installing phpunit/php-timer (2.0.0): Loading from cache
  - Installing sebastian/global-state (2.0.0): Loading from cache
  - Installing sebastian/recursion-context (3.0.0): Loading from cache
  - Installing sebastian/object-reflector (1.1.1): Loading from cache
  - Installing sebastian/object-enumerator (3.0.3): Loading from cache
  - Installing sebastian/resource-operations (1.0.0): Loading from cache
  - Installing myclabs/deep-copy (1.8.1): Loading from cache
  - Installing phar-io/version (2.0.1): Loading from cache
  - Installing phar-io/manifest (1.0.3): Loading from cache
  - Installing doctrine/instantiator (1.1.0): Loading from cache
  - Installing sebastian/exporter (3.1.0): Loading from cache
  - Installing sebastian/diff (3.0.1): Loading from cache
  - Installing sebastian/comparator (3.0.2): Loading from cache
  - Installing webmozart/assert (1.3.0): Loading from cache
  - Installing phpdocumentor/reflection-common (1.0.1): Loading from cache
  - Installing phpdocumentor/type-resolver (0.4.0): Loading from cache
  - Installing phpdocumentor/reflection-docblock (4.3.0): Loading from cache
  - Installing phpspec/prophecy (1.8.0): Loading from cache
sebastian/global-state suggests installing ext-uopz (*)
Writing lock file
Generating optimized autoload files
```
```
$ cp composer.json composer.json.bak
$ composer require --no-update phpunit/php-invoker:^2.0
$ mv composer.json.bak composer.json
```
```
$ composer validate --no-check-lock --strict composer.json
The lock file is not up to date with the latest changes in composer.json, it is recommended that you run `composer update`.
composer.json is valid
$ echo $?
1
```
The exit code of `1` conflicts with the message that says "composer.json is valid".

**Comment by alcohol** [comment]
That depends on how you interpret the message.

`--strict` turns warnings (in this case the hash mismatching between the composer.json and the composer.lock) into errors, thus causing an exit code of 1.

But the `composer.json` is valid in this scenario, as stated.

**Comment by sebastianbergmann** [comment]
How is `--no-check-lock` interpreted then? As I understand it, this should exclude `composer.lock` from being looked at.

**Comment by alcohol** [comment]
Hmm, this option does not behave as documented. Not sure if the behaviour was modified afterwards or if it never had an accurate description. As it is implemented right now though, it will treat the 'out of sync' as a warning _instead of_ an error. But then strict converts the warning back into an error. Ergo, the unexpected exit code. I'll check the commit history to see what happened here.

**Comment by alcohol** [comment]
Yeah, awkward phrasing perhaps in initial implementation by @localheinz https://github.com/composer/composer/commit/5828f6202d332c1c3afe7eb309b3eae994c483a5#diff-959336e114e2ee156fc954e15b4e1faa

**Comment by alcohol** [comment]
The documention entry is more accurate than the flag description in the console command itself.

**Comment by alcohol** [comment]
@sebastianbergmann best I can offer you right now is; if the exit code is 2, there were actual errors. If the exit code equals 1, there were warnings, which normally would result in exit code 0, but due to strict it is elevated to 1.

**Comment by Seldaek** [comment]
Maybe no-check-lock should really not check though.. The other --no-check* options completely skip the checks and not just the warning level. That'd sound like a more reasonable fix to me than tweaking the docs.

**Comment by Seldaek** [comment]
Ah yeah? Ok I thought it skipped it, but then let's keep it as is and fix the docs. Fine by me too.

**Comment by alcohol** [comment]
@Seldaek maybe that used to be the behaviour. I would have to check history again. But currently it merely downgrades errors to warnings.

The `--no-check-all` flag does directly impact the behaviour of the `ConfigValidator` though. So that is the only one that actually skips something I think. But, unlike the name would suggest, it has no impact on the publish or lock checks, so it's not "all" relative to the other flags, but rather the config validator only.

**Comment by yassine-ah** [comment]
@alcohol `outputResult` function pass the args `$errors` and `$warnings` by reference.
I think we can pass them by value to solve the issue.
Also, from what I understand, the current usage (by reference) eliminate our need to check `($publishErrors && $checkPublish) || ($lockErrors && $checkLock)`

**Comment by Seldaek** [comment]
Fixed by #7637
